app/src/main/java/com/rektstudios/trueweather/data/local/RealmDaoImpl.kt

```diff
@@ -1,115 +1,99 @@
+import com.rektstudios.trueweather.domain.util.Constants.MAX_BACKGROUND_COUNT
 import io.realm.Realm
 import io.realm.Sort
 import io.realm.kotlin.toFlow
-import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.asFlow
-import kotlinx.coroutines.flow.flowOn
-import kotlinx.coroutines.withContext
-import java.text.SimpleDateFormat
 import java.util.Calendar
-import java.util.Locale
+import java.util.Random
 import javax.inject.Inject
 
 class RealmDaoImpl @Inject constructor(
     private val realm: Realm
 ):IRealmDao {
-    override suspend fun addCity(cityItem: CityItem)  {
-        withContext(Dispatchers.IO) {
-            realm.copyToRealm(cityItem)
+    override suspend fun addCity(city: String)  {
+        realm.executeTransaction {
+            getCityAsQuery(it, city) ?: run{ it.copyToRealm(CityItem().apply {
+                cityName = city
+                backgroundColor = Random().nextInt(MAX_BACKGROUND_COUNT-1)
+            }) }
         }
     }
 
-    override suspend fun deleteCity(cityItem: CityItem) {
-        withContext(Dispatchers.IO) {
+    override suspend fun deleteCity(city: String) {
             realm.executeTransaction {
-                it.where(CityItem::class.java)
-                    .equalTo("name", cityItem.cityName)
-                    .and()
-                    .equalTo("countryCode", cityItem.countryCode)
-                    .findFirst()
-                    ?.deleteFromRealm()
+                getCityAsQuery(it, city)?.deleteFromRealm()
             }
-        }
     }
 
-    override suspend fun addWeather(cityItem: CityItem, date: String, weatherHourItems: List<WeatherHourItem>) {
-        withContext(Dispatchers.IO) {
-            realm.executeTransaction { iRealm ->
-                val dateItem = queryDateItemHandler(iRealm, cityItem, date)
-                weatherHourItems.forEach { it.dateItem = dateItem }
-                iRealm.copyToRealm(weatherHourItems)
-                dateItem.weatherEveryHour.addAll(weatherHourItems)
-            }
-        }
-    }
+    override suspend fun <T> addWeather(city: String, weather: T) {
+            realm.executeTransaction {
+                getCityAsQuery(it,city)?.let {cityItem ->
+                    when (weather) {
+                        is WeatherHourItem -> {
+                                cityItem.weatherEveryHour
+                                    .where()
+                                    ?.equalTo("timeEpoch", weather.timeEpoch)
+                                    ?.or()
+                                    ?.lessThan("timeEpoch", Calendar.getInstance().timeInMillis/1000L- FORECAST_MIN_TIME_PAST)
+                                    ?.findAll()
+                                    ?.deleteAllFromRealm()
+                            if(weather.timeEpoch>Calendar.getInstance().timeInMillis/1000L- FORECAST_MIN_TIME_PAST)
```

Comment:
please add proper indentation
please add function for Calendar.getInstance().timeInMillis/1000L and reuse